### PR TITLE
zos: add support for IBM Open XL C/C++ & PL/I compilers

### PR DESCRIPTION
**Pull request by gabylb**
##### Checklist
- [x] `npm install && npm test` passes
- [x] commit message follows [commit guidelines](https://github.com/googleapis/release-please#how-should-i-write-my-commits)

##### Description of change
Open XL C/C++ for z/OS (C invocation: clang, ibm-clang, or ibm-clang64) is LLVM-based, and doesn't support XL C/C++'s -q flags. Environment variable CC is set starting with Node.js v18 for z/OS which requires clang or ibm-clang64 compiler, but CC may not be set in older versions (which required njsc or xlclang), hence default to njsc to support older versions.

Also fix defines of `_ALL_SOURCE` and `_UNIX03_SOURCE` to match how they're used.

### CODE REVIEW

File: addon.gypi [hunk 1]
```diff
@@ -109,21 +109,22 @@
...
+              '-qenum=int',
+              '-qxclang=-fexec-charset=ISO8859-1'
+            ],
+          }],
+        ],
+        'ldflags': [
+          '<(node_exp_file)',
         ],
         'defines': [
-          '_ALL_SOURCE=1',
```
Comment (IgorTodorovskiIBM) [inline addon.gypi:120]: Should we leave these defines in, in case the common.gypi is not included (as in earlier editions of v14)?

Comment (gabylb) [reply addon.gypi:120]: Ok, but will remove the `=1` to match those in common.gypi and also how they're used (`#ifdef ...` instead of `#if ...`).

**Review by IgorTodorovskiIBM — APPROVED** [review]
LGTM

File: gyp/pylib/gyp/generator/make.py [hunk 2]
```diff
@@ -103,6 +103,8 @@ def CalculateVariables(default_variables, params):
             default_variables.setdefault("SHARED_LIB_SUFFIX", ".x")
         else:
             default_variables.setdefault("SHARED_LIB_SUFFIX", ".so")
+        if flavor == "zos":
+            COMPILABLE_EXTENSIONS.update({".pli": "pli"})
```
Comment (zsw007) [inline gyp/pylib/gyp/generator/make.py]: Would it make more sense for this to be grouped together with the `elif flavor == "zos":` on line 102-103?

Comment (gabylb) [reply gyp/pylib/gyp/generator/make.py]: Good catch!

File: gyp/pylib/gyp/generator/make.py [hunk 3]
```diff
@@ -2469,16 +2484,27 @@ def CalculateMakefilePath(build_file, base_name):
         header_params.update({"link_commands": LINK_COMMANDS_ANDROID})
     elif flavor == "zos":
         copy_archive_arguments = "-fPR"
-        makedep_arguments = "-qmakedep=gcc"
+        CC = GetEnvironFallback(("CC_target", "CC"), "njsc")
+        if CC == "clang" or CC == "ibm-clang64":
+            makedep_arguments = "-MMD"
+            CC_host = GetEnvironFallback(("CC_host", "CC"), "clang")
+            CXX_target = GetEnvironFallback(("CXX_target", "CXX"), "clang++")
+            CXX_host = GetEnvironFallback(("CXX_host", "CXX"), "clang++")
```
Comment (zsw007) [inline gyp/pylib/gyp/generator/make.py]: If CC is set to `ibm-clang64` and CXX is left unset, then it seems like CXX will still fall back to `clang++`. Could that be a potential issue? Should we make sure it fallbacks to `ibm-clang++64` instead?

File: gyp/pylib/gyp/generator/make.py [hunk 4]
```diff
@@ -2469,16 +2483,34 @@ def CalculateMakefilePath(build_file, base_name):
         header_params.update({"link_commands": LINK_COMMANDS_ANDROID})
     elif flavor == "zos":
         copy_archive_arguments = "-fPR"
-        makedep_arguments = "-qmakedep=gcc"
+        CC = GetEnvironFallback(("CC_target", "CC"), "njsc")
+        makedep_arguments = "-MMD"
+        if CC == "clang":
+            CC_host = CC
+            CC_host = GetEnvironFallback(("CC_host", "CC"), "clang")
```
Comment (zsw007) [inline gyp/pylib/gyp/generator/make.py]: It looks like you assign `CC_host` twice here, is this intended?

File: gyp/pylib/gyp/generator/make.py [hunk 5]
```diff
@@ -2469,16 +2483,34 @@ def CalculateMakefilePath(build_file, base_name):
...
+            CXX_target = GetEnvironFallback(("CXX_target", "CXX"), "clang++")
+            CXX_host = GetEnvironFallback(("CXX_host", "CXX"), "clang++")
+        elif CC == "ibm-clang64":
+            CC_host = CC
+            CC_host = GetEnvironFallback(("CC_host", "CC"), "ibm-clang64")
```
Comment (zsw007) [inline gyp/pylib/gyp/generator/make.py]: It looks like `CC_host` is assigned twice here as well similar to https://github.com/ibmruntimes/node-gyp/pull/5/files#r982693141

**Review by zsw007 — APPROVED** [review]
LGTM

**Review by joransiu — APPROVED** [review]
LGTM.   Do we need to handle the case if someone sets CC to `ibm-clang`?  Or is that simply not supported, and users should use `ibm-clang64`?

**Comment by gabylb** [comment]
> LGTM. Do we need to handle the case if someone sets CC to `ibm-clang`? Or is that simply not supported, and users should use `ibm-clang64`?

In the Node.js v18+ setup we'll include setting of CC, CXX and LINK in a `.env`, and they'll all be `clang|clang++`. But you're right we should support it. Thanks.

**Review by joransiu — APPROVED** [review]

**Review by IgorTodorovskiIBM — APPROVED** [review]
LGTM

**Review by zsw007 — APPROVED** [review]
LGTM

**Review by mrjolin — APPROVED** [review]
I approve upstreaming these changes.

**Pull request merged**